### PR TITLE
NUI-884 example worker that calls azure api's

### PR DESCRIPTION
**Pull request by jdelbick**
jira ticket: https://jira.corp.adobe.com/browse/NUI-884

Changes:
- moved over base content from the intern Justin's work this past summer: https://git.corp.adobe.com/calloway/azure_worker
- cleaned up Justin's code and fixed the tests
- also removed the `threshold` param on the pdfextract API (requested by the Sensei team to do this)

TODO: 
- decide if we want these workers to produce metadata renditions instead of `.json`

### CODE REVIEW

**Comment by jdelbick** [comment]
@chandarn added you to this so you can help answer my question, should we convert these workers into metadata renditions? (currently they produce json files)

File: projects/worker-azure-ai/.gitignore [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+jspm_packages
+
+# build
+build
+dist
+.manifest-dist.yml
+
+# Config
+config.json
+.env*
```
Comment (tmathern) [inline projects/worker-azure-ai/.gitignore:13]: Typo:
```suggestion
*.env
```

File: projects/worker-azure-ai/actions/worker-azure-ocr/index.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        }),
+        headers: {
+            "Content-Type": "application/json",
+            "Ocp-Apim-Subscription-Key" : subscriptionKey
+        }
+    };
+
+    const response = await fetch(url, options);
+    const jsonResponse = await response.json();
+    if (jsonResponse.code || jsonResponse.error) {
```
Comment (tmathern) [inline projects/worker-azure-ai/actions/worker-azure-ocr/index.js:43]: So there is a `jsonResponse.code` only on failure, not on success ('20*')?

File: projects/worker-azure-ai/actions/worker-azure-ocr/index.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    let subscriptionKey = params && params.AZURE_OCP_KEY;
+    let endpoint = params && params.AZURE_OCP_ENDPOINT;
+
+    // WORKER_TEST_MODE is true when you are running worker unit
+    // tests using the `test-worker` command
+    // Dummy authorization parameters needed for mocks 
+    if (process.env.WORKER_TEST_MODE) {
+        subscriptionKey = "test-azure-key";
+        endpoint = "https://westus.api.cognitive.microsoft.com/";
+    }
```
Comment (tmathern) [inline projects/worker-azure-ai/actions/worker-azure-ocr/index.js:20]: I would suggest `if/else` here with below condition. I nearly missed the below condition and asked to check it after assignments! ^^'

File: projects/worker-azure-ai/actions/worker-azure-tagging/index.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    let subscriptionKey = params && params.AZURE_OCP_KEY;
+    let endpoint = params && params.AZURE_OCP_ENDPOINT;
+
+    // WORKER_TEST_MODE is true when you are running worker unit
+    // tests using the `test-worker` command
+    // Dummy authorization parameters needed for mocks 
+    if (process.env.WORKER_TEST_MODE) {
+        subscriptionKey = "test-azure-key";
+        endpoint = "https://westus.api.cognitive.microsoft.com/";
+    }
```
Comment (tmathern) [inline projects/worker-azure-ai/actions/worker-azure-tagging/index.js:20]: I would suggest `if/else` here with below condition. I nearly missed the below condition and asked to check it after assignments! ^^'

File: projects/worker-azure-ai/actions/worker-azure-tagging/index.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        }),
+        headers: {
+            "Content-Type": "application/json",
+            "Ocp-Apim-Subscription-Key" : subscriptionKey
+        }
+    };
+
+    const response = await fetch(url, options);
+    const jsonResponse = await response.json();
+    if (jsonResponse.code || jsonResponse.error) {
```
Comment (tmathern) [inline projects/worker-azure-ai/actions/worker-azure-tagging/index.js:45]: So there is a jsonResponse.code only on failure, not on success ('20*')?

File: projects/worker-azure-ai/package.json [hunk 6]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "azure-worker-ai",
+  "version": "0.0.1",
+  "private": true,
+  "dependencies": {
+    "@adobe/asset-compute-sdk": "^2.2.1",
```
Comment (tmathern) [inline projects/worker-azure-ai/package.json:6]: Update the SDK to latest?
(Check post-processing won't run and fail before updating).

File: projects/worker-azure-ai/test/asset-compute/worker-azure-ocr/invalid-language/params.json [hunk 7]
```diff
@@ -0,0 +1,5 @@
+{
+  "errorReason": "GenericError",
+  "language": "garbage",
```
Comment (tmathern) [inline projects/worker-azure-ai/test/asset-compute/worker-azure-ocr/invalid-language/params.json:3]: Sounds better:
```suggestion
  "language": "unknown-language",
```

File: projects/worker-azure-ai/test/asset-compute/worker-azure-tagging/invalid-language/params.json [hunk 8]
```diff
@@ -0,0 +1,5 @@
+{
+  "errorReason": "GenericError",
+  "language": "garbage",
```
Comment (tmathern) [inline projects/worker-azure-ai/test/asset-compute/worker-azure-tagging/invalid-language/params.json:3]:
```suggestion
  "language": "unknown-language",
```

**Review by tmathern — CHANGES_REQUESTED** [review]

File: projects/worker-azure-ai/actions/worker-azure-ocr/index.js (hunk 2 above)
Comment (jdelbick) [reply projects/worker-azure-ai/actions/worker-azure-ocr/index.js:43]: the code is not a number, it's a string value defined by the azure api: https://westcentralus.dev.cognitive.microsoft.com/docs/services/5adf991815e1060e6355ad44/operations/56f91f2e778daf14a499e1fc
```
{
    "code":"InvalidImageFormat",
    "requestId":"B8D802CF-DD8F-4E61-B15C-9E6C5844CCBC",
    "message":"The input file is not in a valid image format that the service can support. "
}
```

File: projects/worker-azure-ai/actions/worker-azure-ocr/index.js (hunk 3 above)
Comment (jdelbick) [reply projects/worker-azure-ai/actions/worker-azure-ocr/index.js:20]: I mean its not really an if/else situation. this is a separate check (and could even be put in a separate function as Mark did in his custom worker)

File: projects/worker-azure-ai/package.json (hunk 6 above)
Comment (jdelbick) [reply projects/worker-azure-ai/package.json:6]: we have to update our generator :)

File: projects/worker-azure-ai/actions/worker-azure-ocr/index.js (hunk 3 above)
Comment (jdelbick) [reply projects/worker-azure-ai/actions/worker-azure-ocr/index.js:20]: to simplify the code, I just removed the credentials check. We can let the api call handle the error.

Comment (tmathern) [reply projects/worker-azure-ai/actions/worker-azure-ocr/index.js:20]: For the record: I hope credentials you set for tests do not need verification, so it is an if/else situation.

File: projects/worker-azure-ai/package.json [hunk 9]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "azure-worker-ai",
+  "version": "0.0.1",
```
Comment (tmathern) [inline projects/worker-azure-ai/package.json:3]:
```suggestion
  "version": "0.1.0",
```
(1.0.0 would be fine too).

**Review by tmathern — APPROVED** [review]

**Comment by alexkli** [comment]
@bobvanmanen what about @jdelbick's question above about output format json vs xmp?

IMO we can stick to json as currently AEM has no generic way to read such machine metadata and it should probably not blindly be mixed inside jcr:content/metadata (assuming we make this an xmp custom metadata style format).

**Comment by chandarn** [comment]
> @bobvanmanen what about @jdelbick's question above about output format json vs xmp?
> 
> IMO we can stick to json as currently AEM has no generic way to read such machine metadata and it should probably not blindly be mixed inside jcr:content/metadata (assuming we make this an xmp custom metadata style format).

+1. We already have an example for metadata rendition. This example is to show the support for an external API/service integration.

**Comment by bobvanmanen** [comment]
@alexkli @chandarn - I agree. The problem, which also is true for the color extraction worker, is that machine generated metadata is not just a flat list of values which means it'll require an AEM customization to view them properly. Also, mixing machine generated metadata, with user provided metadata, and metadata extracted from the original file is not trivial.

**Comment by jdelbick** [comment]
@bobvanmanen to make sure I understand correctly what you are saying:
OCR -  leave it as JSON
tagging - IIUC, same as ccai workers: write to xmp using [custom namespace](https://github.com/adobe/asset-compute-example-workers/blob/master/projects/worker-ccai-pdfextract/worker-ccai-pdfextract.js#L25), list instructions to [create custom metadata schema](https://github.com/adobe/asset-compute-example-workers/tree/master/projects/worker-ccai-pdfextract#update-metadata-schema) 
regarding the metadata schema, maybe something like `azureTags:`?

**Review by DrGirlfriend — APPROVED** [review]

**Comment by jdelbick** [comment]
@alexkli @chandarn @bobvanmanen when you have a moment, I'd like some feedback on potential ideas for metadata schema for azure smart tags

if we think this example is already covered enough in the [ccai workers](https://github.com/adobe/asset-compute-example-workers/blob/master/projects/worker-ccai-colorextract/worker-ccai-colorextract.js#L172-L186) then I can just merge this code as is -- with JSON renditions

File: projects/worker-azure-ai/actions/worker-azure-tagging/index.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
+"use strict";
+
+const { worker, GenericError } = require("@adobe/asset-compute-sdk");
+const fetch = require("node-fetch");
+const fs = require('fs').promises;
+
+const DEFAULT_LANGUAGE = "en";
+
+exports.main = worker(async (source, rendition, params) => {
```
Comment (alexkli) [inline projects/worker-azure-ai/actions/worker-azure-tagging/index.js:9]: Should this get `disableSourceDownload: true` as well?

File: projects/worker-azure-ai/actions/worker-azure-ocr/index.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    const jsonResponse = await response.json();
+    if (jsonResponse.code || jsonResponse.error) {  
+        // Format error code and message like the following:
+        // `The Azure OCR API failed with FailedToProcess: Could not extract image features`
+        throw new GenericError(`The Azure OCR Api failed with ${jsonResponse.code}: ${jsonResponse.error}`);
+    }
+
+    await fs.writeFile(rendition.path, JSON.stringify(jsonResponse));
+}, {
+    disableSourceDownload:"true" // source presigned url is passed directly to azure api
```
Comment (alexkli) [inline projects/worker-azure-ai/actions/worker-azure-ocr/index.js:46]: Should be boolean `true`.

File: projects/worker-azure-ai/package.json (hunk 9 above)
Comment (alexkli) [reply projects/worker-azure-ai/package.json:3]: I guess use `1.0.0` consistent with [other example](https://github.com/adobe/asset-compute-example-workers/blob/master/projects/worker-metadata/package.json#L4) workers.

**Review by alexkli — APPROVED** [review]
I leave the output format to the AEM team. From a Nui perspective this looks good as an example.

See small requests inline.

**Pull request merged**